### DISCUSSION TITLE
ErrNoError on TopicCreate

### DISCUSSION THREAD

**Discussion by maciej** [IBM/sarama#2766] [General]
The call to CreateTopics using `sarama.Broker` returns a `TopicErrors` map containing errors on topic creation. However it does also return a quite peculiar error:

```
go
const (
	ErrNoError                            KError = 0
// ...
)
```

Why was the API designed this way? It seems not to be idiomatic Go.

**Comment by dnwe** [comment]
@maciej 👋🏻  thanks for getting in touch

Those 'KError' codes map to the return codes that the Kafka protocol returns.

So CreateTopics on sarama.Broker (although an exported name) isn't really intended for Sarama users, but is roundtripping the [CreateTopics protocol message](https://kafka.apache.org/protocol.html#The_Messages_CreateTopics) to Kafka and receiving the response in a CreateTopicsResponse struct
https://github.com/IBM/sarama/blob/c067a7f4e5653479b039a09f9978a28a54584e54/broker.go#L629-L638

Typically a Sarama user would be creating topics through the Admin client interface, which _should_ return more idiomatic Go types and expose less of the internals of the client (although we have a backlog item to improve that separation in /v2 of Sarama
https://github.com/IBM/sarama/blob/c067a7f4e5653479b039a09f9978a28a54584e54/admin.go#L223

**Comment by maciej** [comment]
Thanks for getting back to me quickly on that @dnwe ! I have missed `ClusterAdmin` when browsing the library interfaces. Thanks for pointing me to it.

With my question answered I'm considering the discussion closed.

**Reply by dnwe** [reply]
No worries at all, feel free to post any questions you have as these are always handy to help buildup a FAQ that others users can find in the future
